### PR TITLE
Role escalation prevention

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR...

Fixes: #5781

Doc PR: https://github.com/apache/cloudstack-documentation/pull/260

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
Thanks @DaanHoogland, I'll review and get back in a few days. Meanwhile, do add/ask other devs.

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/user/AccountManagerImpl.java:1067]:
```suggestion
        // Check role escalation permission for the account
```

remove comment if not required

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (Pearl1594) [inline server/src/main/java/com/cloud/user/AccountManagerImpl.java:1140]: maybe this if-else clause isn't required?

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/user/AccountManagerImpl.java:1140]: :) it will be when done, still in progress @Pearl1594

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/user/AccountManagerImpl.java:1067]: work in progress @sureshanaparti . I need to keep options open atm.

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: api/src/test/java/org/apache/cloudstack/api/command/admin/account/CreateAccountCmdTest.java
Comment (rohityadavcloud) [inline api/src/test/java/org/apache/cloudstack/api/command/admin/account/CreateAccountCmdTest.java:76]: in setUp(), should we set the roleId field too?

File: plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java
Comment (rohityadavcloud) [inline plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java:108]: Could this cause any issue, cc @Pearl1594 ?

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (rohityadavcloud) [inline server/src/main/java/com/cloud/user/AccountManagerImpl.java:398]: Should we only get api name list from the configured pluggable service (the dynamic or static manager?)

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (rohityadavcloud) [inline server/src/main/java/com/cloud/user/AccountManagerImpl.java:413]: Can we re-use parts of this from api discovery manager/service?

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (rohityadavcloud) [inline server/src/main/java/com/cloud/user/AccountManagerImpl.java:1192]: Could there be a simpler fix, that is the caller is only allowed to select/pass roleids while creating account that is equal or lower than their own role; for example;
root admin -> can do anything
domain admin -> create domain/sub-domain admins and user-accounts
user accounts -> none

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
Most / all test failures report the following error: `Execute cmd: createaccount failed, due to: errorCode: 531, errorText:user of account admin/1 (5af79319-7414-4df3-bffb-ec4ef9728066) can not create an account with access to notifyBaremetalProvisionDone\n']` and the `notifyBaremetalProvisionDone` is only allowed for an account having User role.

**Comment by DaanHoogland** [comment]
> Most / all test failures report the following error: `Execute cmd: createaccount failed, due to: errorCode: 531, errorText:user of account admin/1 (5af79319-7414-4df3-bffb-ec4ef9728066) can not create an account with access to notifyBaremetalProvisionDone\n']` and the `notifyBaremetalProvisionDone` is only allowed for an account having User role.

ok, that is not good: `description = "Notify provision has been done on a host. This api is for baremetal virtual router service, not for end user"`, it is hard-coded to be only allowed for the role user but is not allowed for normal users. Also I tested creating accounts as domain admin. I wonder why I didn't notice this one!?

I removed the restriction to RoleTupe.User, which should do the trick. If not I'll add all role types.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/user/AccountManagerImpl.java:1192]: that would not be taking into account dynamic roles. We want the caller not to be able to create an account with a role that has any permission that they don't have themselves, independent of the RoleType.

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/user/AccountManagerImpl.java:413]: I'll look into it again. It didn't seem prudent when I implemented this.

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/user/AccountManagerImpl.java:398]: this implementation is cleverly stolen. I'll have a look.

File: api/src/test/java/org/apache/cloudstack/api/command/admin/account/CreateAccountCmdTest.java
Comment (DaanHoogland) [reply api/src/test/java/org/apache/cloudstack/api/command/admin/account/CreateAccountCmdTest.java:76]: makes sense, will do

Comment (DaanHoogland) [reply api/src/test/java/org/apache/cloudstack/api/command/admin/account/CreateAccountCmdTest.java:76]: It actually doesn't matter for this flimsy test case. We are only verifying that the command is passed to the service, not the handling of it in the service. The service is a mock.

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/user/AccountManagerImpl.java:398]: No, we need the full API list by name, the plugable service hold class references, which is overkill for us at this moment.

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/user/AccountManagerImpl.java:398]: the role service can be dis-/enabled dynamically. we need a conclusive list of all API. It needs to be done independent of which checker is active.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
ping @DaanHoogland @sureshanaparti @nvazquez  - is this ready to go or need a set of thorough testing and review?

**Comment by DaanHoogland** [comment]
@rohityadavcloud , It needs a second review (I think @sureshanaparti tested)

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/user/AccountManagerImpl.java:1222]: @DaanHoogland Update the message as appropriate (do not include any API name which the user doesn't have access to)

File: plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java
Comment (nvazquez) [inline plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java:66]: Not an issue but an improvement request - can we leave only one of these methods?

File: plugins/api/discovery/src/main/java/org/apache/cloudstack/discovery/ApiDiscoveryServiceImpl.java
Comment (nvazquez) [inline plugins/api/discovery/src/main/java/org/apache/cloudstack/discovery/ApiDiscoveryServiceImpl.java:223]: I think we won't need this variable - after the checkAccess method we can add the element to the list and on the catch block we can log the exception describing the error. Then the next if won't be needed either

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (nvazquez) [inline server/src/main/java/com/cloud/user/AccountManagerImpl.java:394]: Could somehow the list be empty but not null at this point?

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (nvazquez) [inline server/src/main/java/com/cloud/user/AccountManagerImpl.java:1242]: Typo?

File: plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java
Comment (DaanHoogland) [reply plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java:66]: both are used, so no ;)

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/user/AccountManagerImpl.java:1242]: yes

File: plugins/api/discovery/src/main/java/org/apache/cloudstack/discovery/ApiDiscoveryServiceImpl.java
Comment (DaanHoogland) [reply plugins/api/discovery/src/main/java/org/apache/cloudstack/discovery/ApiDiscoveryServiceImpl.java:223]: no, multiple access checkers can be called and any one may deny.

File: server/src/main/java/com/cloud/user/AccountManagerImpl.java
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/user/AccountManagerImpl.java:394]: no, the private apinamelist is not initialised till in this block and nowhere else.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
cc @sureshanaparti probably need to investigate why deb builds are failing
@blueorangutan test

**Comment by DaanHoogland** [comment]
> cc @sureshanaparti probably need to investigate why deb builds are failing @blueorangutan test

three days ago deb builds didn't fail. I think it is a environmental fluke,

**Comment by sureshanaparti** [comment]
ping @rohityadavcloud can you test this changes.

**Comment by rohityadavcloud** [comment]
@sureshanaparti I'll review the code changes but not able to test it. Pl ask @borisstoyanov @vladimirpetrov or one of the 4.16.1 team.

**Comment by sureshanaparti** [comment]
> @sureshanaparti I'll review the code changes but not able to test it. Pl ask @borisstoyanov @vladimirpetrov or one of the 4.16.1 team.

sure @rohityadavcloud , please review the changes. thanks.

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually tested, restricted domain admin cannot create a role with higher privileges

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
Do we need this re-test 

> @sureshanaparti a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

I think the one before showed no errors and we didn't re-build in between.

**Comment by sureshanaparti** [comment]
> Do we need this re-test
> 
> > @sureshanaparti a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests
> 
> I think the one before showed no errors and we didn't re-build in between.

couldn't see the results here, let's wait for test results.
https://github.com/apache/cloudstack/pull/5879#issuecomment-1033261423

File: plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java
Comment (Pearl1594) [reply plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java:112]: This doesn't seem to affect the project roles feature.

**Review by Pearl1594 — APPROVED** [review]
LGTM - verified behavior.

File: plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java
Comment (sureshanaparti) [reply plugins/acl/project-role-based/src/main/java/org/apache/cloudstack/acl/ProjectRoleBasedApiAccessChecker.java:112]: thanks for confirming this @Pearl1594

**Review by rohityadavcloud — COMMENTED** [review]
FWIW - LGTM, read the critical code in checkRoleEscalation, will leave some remarks

**Comment by rohityadavcloud** [comment]
@DaanHoogland one remark this doesn't fix/address if the issue already exists, we should update the doc (if not already done) to explain how roles work and precautions for non-admin roles that can create accounts.

**Comment by DaanHoogland** [comment]
> @DaanHoogland one remark this doesn't fix/address if the issue already exists, we should update the doc (if not already done) to explain how roles work and precautions for non-admin roles that can create accounts.

I don't agree, there is a doc PR out. Please comment there if there is anything missing.

**Comment by sureshanaparti** [comment]
> > @DaanHoogland one remark this doesn't fix/address if the issue already exists, we should update the doc (if not already done) to explain how roles work and precautions for non-admin roles that can create accounts.
> 
> I don't agree, there is a doc PR out. Please comment there if there is anything missing.

@rohityadavcloud Doc PR here, https://github.com/apache/cloudstack-documentation/pull/260, please take a look and provide your feedback. @DaanHoogland can create a separate doc PR to include if any more details has to be added to the documentation.

**Comment by rohityadavcloud** [comment]
Cool thanks for the link @sureshanaparti @DaanHoogland - as I said my remark was only if you've not already done the doc PR - which you've.

**Pull request merged**
